target/ppc/int_helper.c

```diff
@@ -1528,6 +1528,59 @@ VEXTU_X_DO(vextuhrx, 16, 0)
+    int array_size = ARRAY_SIZE(t->u8);                                     \
+    ppc_avr_t tmp = { .u64 = { 0, 0 } };                                    \
+    ofs = ofs & 0x1F;                                                       \
+                                                                            \
+    if (ofs + sz <= array_size) {                                           \
+        memmove(&tmp.u8[array_size / 2 - sz], &a->u8[ofs], sz);             \
+    } else if (ofs >= array_size) {                                         \
+        memmove(&tmp.u8[array_size / 2 - sz], &b->u8[ofs - array_size], sz);\
+    } else {                                                                \
+        int len_a = array_size - ofs, len_b = sz - len_a;                   \
+                                                                            \
+        memmove(&tmp.u8[array_size / 2 - sz], &a->u8[array_size - len_a],   \
+                len_a);                                                     \
+        memmove(&tmp.u8[array_size / 2 - len_b], &b->u8[0], len_b);         \
+    }                                                                       \
+                                                                            \
+    *t = tmp;                                                               \
+}
+#else
+#define VEXTDVLX(name, sz) \
+void glue(helper_, name)(ppc_avr_t *t, ppc_avr_t *a, ppc_avr_t *b,          \
+                         target_ulong ofs)                                  \
+{                                                                           \
+    int array_size = ARRAY_SIZE(t->u8);                                     \
+    ppc_avr_t tmp = { .u64 = { 0, 0 } };                                    \
+    ofs = ofs & 0x1F;                                                       \
+                                                                            \
+    if (ofs + sz <= array_size) {                                           \
+        memmove(&tmp.u8[array_size / 2], &a->u8[array_size - sz - ofs], sz);\
+    } else if (ofs >= array_size) {                                         \
+        ofs -= array_size;                                                  \
+        memmove(&tmp.u8[array_size / 2], &b->u8[array_size - sz - ofs], sz);\
+    } else {                                                                \
+        int len_a = array_size - ofs, len_b = sz - len_a;                   \
+                                                                            \
+        memmove(&tmp.u8[array_size / 2], &b->u8[array_size - len_b], len_b);\
+        memmove(&tmp.u8[array_size / 2 + len_b], &a->u8[0], len_a);         \
+    }                                                                       \
+                                                                            \
+    *t = tmp;                                                               \
+}
+#endif
+VEXTDVLX(vextdubvlx, 1)
```

Comment:
I tried to apply some of this in https://github.com/PPC64/qemu/pull/25/commits/228f6aa37ea8d9afd7e891345ae5b769c2443ebf
- Checks within the helper, and also log a guest error for invalid indexes
- I prefer to keep all cases in the "main" macro, handling the non-crossing case in a ternary operator seems less readable
- It seems easier to pass the type and use `sizeof ` instead of the size in bits